File: packages/client/src/process/spawn.rs
use {
	crate::prelude::*,
	futures::{Stream, TryStreamExt as _, future},
	serde_with::serde_as,
	tangram_http::{request::builder::Ext as _, response::Ext as _},
	tangram_util::serde::{CommaSeparatedString, is_false},
};

#[serde_as]
#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
pub struct Arg {
	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub cached: Option<bool>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub checksum: Option<tg::Checksum>,

	pub command: tg::Referent<tg::command::Id>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub local: Option<bool>,

	#[serde(default, skip_serializing_if = "Vec::is_empty")]
	pub mounts: Vec<tg::process::data::Mount>,

	#[serde(default, skip_serializing_if = "is_false")]
	pub network: bool,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub parent: Option<tg::process::Id>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[serde_as(as = "Option<CommaSeparatedString>")]
	pub remotes: Option<Vec<String>>,

	#[serde(default, skip_serializing_if = "is_false")]
	pub retry: bool,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub stderr: Option<tg::process::Stdio>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub stdin: Option<tg::process::Stdio>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub stdout: Option<tg::process::Stdio>,
}

#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
pub struct Output {
	pub process: tg::process::Id,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub remote: Option<String>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub token: Option<String>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	pub wait: Option<tg::process::wait::Output>,
}

impl tg::Client {
	pub async fn try_spawn_process(
		&self,
		arg: tg::process::spawn::Arg,
	) -> tg::Result<
		impl Stream<Item = tg::Result<tg::progress::Event<Option<tg::process::spawn::Output>>>>
		+ Send
		+ 'static,
	> {
		let method = http::Method::POST;
		let uri = "/processes/spawn";
		let request = http::request::Builder::default()
			.method(method)
			.uri(uri)
			.header(http::header::ACCEPT, mime::TEXT_EVENT_STREAM.to_string())
			.header(
				http::header::CONTENT_TYPE,
				mime::APPLICATION_JSON.to_string(),
			)
			.json(arg)
			.map_err(|source| tg::error!(!source, "failed to serialize the arg"))?
			.unwrap();
		let response = self
			.send_with_retry(request)
			.await
			.map_err(|source| tg::error!(!source, "failed to send the request"))?;
		if !response.status().is_success() {
			let error = response.json().await.map_err(|source| {
				tg::error!(!source, "failed to deserialize the error response")
			})?;
			return Err(error);
		}
		let stream = response
			.sse()
			.map_err(|source| tg::error!(!source, "failed to read an event"))
			.and_then(|event| {
				future::ready(
					if event.event.as_deref().is_some_and(|event| event == "error") {
						match event.try_into() {
							Ok(error) | Err(error) => Err(error),
						}
					} else {
						event.try_into()
					},
				)
			});
		Ok(stream)
	}
}

impl Arg {
	#[must_use]
	pub fn with_command(command: tg::Referent<tg::command::Id>) -> Self {
		Self {
			cached: None,
			checksum: None,
			command,
			local: None,
			mounts: Vec::new(),
			network: false,
			parent: None,
			remotes: None,
			retry: false,
			stderr: None,
			stdin: None,
			stdout: None,
		}
	}

	#[must_use]
	pub fn with_command_and_checksum(
		command: tg::Referent<tg::command::Id>,
		checksum: Option<tg::Checksum>,
	) -> Self {
		Self {
			cached: None,
			checksum,
			command,
			local: None,
			mounts: Vec::new(),
			network: false,
			parent: None,
			remotes: None,
			retry: false,
			stderr: None,
			stdin: None,
			stdout: None,
		}
	}
}

File: packages/clients/js/src/client/process/spawn.ts
import * as tg from "../../index.ts";
import { Body, Request } from "../../http.ts";
import type { Client } from "../../client.ts";

export namespace Spawn {
	export type Arg = {
		cached?: boolean;
		cacheLocation?: tg.Location.Arg | null;
		checksum?: tg.Checksum | null;
		command: tg.Referent<tg.Command.Id>;
		debug?: tg.Process.Debug | null;
		location?: tg.Location.Arg | null;
		parent?: tg.Process.Id | null;
		public?: boolean;
		retry?: boolean;
		sandbox?: tg.Sandbox.DataArg | string | null;
		stderr?: tg.Process.Stdio;
		stdin?: tg.Process.Stdio;
		stdout?: tg.Process.Stdio;
		tty?: boolean | tg.Process.Tty | null;
	};

	export namespace Arg {
		export let toJson = (arg: tg.Process.Spawn.Arg): unknown => {
			let output: { [key: string]: unknown } = {};
			if (arg.cached !== undefined) {
				output.cached = arg.cached;
			}
			if (arg.cacheLocation !== undefined) {
				output.cache_location =
					arg.cacheLocation === null
						? null
						: tg.Location.Arg.toDataString(arg.cacheLocation);
			}
			if (arg.checksum !== undefined) {
				output.checksum = arg.checksum;
			}
			output.command = tg.Referent.toData(arg.command, (id) => id);
			if (arg.debug !== undefined) {
				output.debug = arg.debug;
			}
			if (arg.location !== undefined) {
				output.location =
					arg.location === null
						? null
						: tg.Location.Arg.toDataString(arg.location);
			}
			if (arg.parent !== undefined) {
				output.parent = arg.parent;
			}
			if (arg.public) {
				output.public = arg.public;
			}
			if (arg.retry) {
				output.retry = arg.retry;
			}
			if (arg.sandbox !== undefined) {
				output.sandbox = arg.sandbox;
			}
			if (arg.stderr !== undefined && arg.stderr !== "inherit") {
				output.stderr = arg.stderr;
			}
			if (arg.stdin !== undefined && arg.stdin !== "inherit") {
				output.stdin = arg.stdin;
			}
			if (arg.stdout !== undefined && arg.stdout !== "inherit") {
				output.stdout = arg.stdout;
			}
			if (arg.tty !== undefined) {
				output.tty = arg.tty;
			}
			return output;
		};
	}

	export type Output = {
		cached: boolean;
		lease?: string | null;
		location?: tg.Location | null;
		process: number | tg.Process.Id;
		token?: tg.Grant.Token | null;
		wait?: tg.Process.Wait.Data | null;
	};

	export namespace Output {
		export let fromJson = (json: unknown): tg.Process.Spawn.Output => {
			let output = json as Omit<tg.Process.Spawn.Output, "location"> & {
				location?: string | tg.Location;
			};
			let { location, wait, ...rest } = output;
			let result: tg.Process.Spawn.Output = { ...rest };
			if (location !== undefined) {
				result.location =
					typeof location === "string"
						? tg.Location.fromDataString(location)
						: location;
			}
			if (wait !== undefined) {
				result.wait = wait;
			}
			return result;
		};
	}
}

export async function spawnProcess(
	client: Client,
	arg: tg.Process.Spawn.Arg,
): Promise<AsyncIterableIterator<tg.Progress.Event<tg.Process.Spawn.Output>>> {
	let stream = await trySpawnProcess(client, arg);
	return mapSpawnEvents(stream);
}

export async function trySpawnProcess(
	client: Client,
	arg: tg.Process.Spawn.Arg,
): Promise<
	AsyncIterableIterator<tg.Progress.Event<tg.Process.Spawn.Output | null>>
> {
	let method = "POST";
	let uri = "/processes/spawn";
	let headers = {
		accept: "text/event-stream",
		"content-type": "application/json",
	};
	let body = Body.json(Spawn.Arg.toJson(arg));
	let request = new Request({
		body,
		method,
		uri,
		headers,
	});
	let response = await client.send(request);
	if (response.status < 200 || response.status >= 300) {
		throw tg.Error.fromData(await response.json<tg.Error.Data>());
	}
	return tg.Progress.decode<tg.Process.Spawn.Output | null>(
		response,
		(output) =>
			output === undefined || output === null
				? null
				: tg.Process.Spawn.Output.fromJson(output),
	);
}

async function* mapSpawnEvents(
	events: AsyncIterable<tg.Progress.Event<tg.Process.Spawn.Output | null>>,
): AsyncIterableIterator<tg.Progress.Event<tg.Process.Spawn.Output>> {
	for await (let event of events) {
		if (event.kind === "output") {
			if (event.value === null) {
				throw new Error("expected a process");
			}
			yield event as tg.Progress.Event<tg.Process.Spawn.Output>;
		} else {
			yield event;
		}
	}
}
